File: whiteboard-app/src/test/java/dev/oscarrojas/whiteboard/ws/protocol/AppEventBinaryConverterTest.java
package dev.oscarrojas.whiteboard.ws.protocol;

import dev.oscarrojas.whiteboard.messaging.AppEvent;
import org.junit.jupiter.api.Test;

import java.nio.ByteBuffer;
import java.nio.charset.StandardCharsets;

import static org.junit.jupiter.api.Assertions.assertEquals;

public class AppEventBinaryConverterTest {

    @Test
    void toBytes() {
        String name = "eventName";

        byte[] payload = new byte[8];
        payload[0] = (byte) 219;
        payload[1] = (byte) 109;
        payload[2] = (byte) 182;
        payload[3] = (byte) 219;
        payload[4] = (byte) 219;
        payload[5] = (byte) 109;
        payload[6] = (byte) 182;
        payload[7] = (byte) 219;
        AppEventBinaryConverter converter = new AppEventBinaryConverter();
        AppEvent event = new AppEvent(name, payload);

        byte[] result = converter.toBytes(event);
        ByteBuffer buffer = ByteBuffer.wrap(result);

        for (byte b : result) {
            System.out.print(b + " ");
        }

        // epoch milliseconds equals timestamp epoch milliseconds
        long milliseconds = buffer.getLong();
        assertEquals(milliseconds, event.getTimestamp().toEpochMilli());

        // name header equals name byte length
        int nameByteLength = buffer.get() & 0xFF;
        assertEquals(
            name.getBytes(StandardCharsets.UTF_8).length, nameByteLength);

        // name bytes equals event name
        byte[] nameBytes = new byte[nameByteLength];
        buffer.get(nameBytes, 0, nameByteLength);
        assertEquals(name, new String(nameBytes, StandardCharsets.UTF_8));

        // payload header equals payload byte length
        int payloadByteLength = buffer.getInt();
        assertEquals(payload.length, payloadByteLength);

        // payload bytes equals message payload
        byte[] resultPayload = new byte[payload.length];
        buffer.get(resultPayload, 0, payload.length);
        for (int i = 0; i < payload.length; i++) {
            assertEquals(payload[i], resultPayload[i]);
        }

        // ensure no remaining bytes are present
        assertEquals(0, buffer.remaining());
    }

    @Test
    void fromBytes() throws BinaryDecodingException {
        String name = "eventName";
        byte[] payload = new byte[8];
        payload[0] = (byte) 219;
        payload[1] = (byte) 109;
        payload[2] = (byte) 182;
        payload[3] = (byte) 219;
        payload[4] = (byte) 219;
        payload[5] = (byte) 109;
        payload[6] = (byte) 182;
        payload[7] = (byte) 219;
        AppEventBinaryConverter converter = new AppEventBinaryConverter();
        AppEvent event = new AppEvent(name, payload);
        byte[] eventBytes = converter.toBytes(event);

        AppEvent result = converter.fromBytes(eventBytes);

        // decoded timestamp equals original timestamp
        assertEquals(event.getTimestamp(), result.getTimestamp());

        // decoded name equals original name
        assertEquals(name, result.getName());

        // decoded payload equals original payload
        byte[] decodedPayload = result.getPayload();
        for (int i = 0; i < payload.length; i++) {
            assertEquals(payload[i], decodedPayload[i]);
        }
    }
}

File: client/src/messaging/__tests__/AppEventBinaryConverter.test.ts
import { test, expect } from "vitest";
import AppEventBinaryConverter, {
  NAME_HEADER_BYTES,
  EPOCH_MILLISECOND_BYTES,
  PAYLOAD_HEADER_BYTES,
} from "../AppEventBinaryConverter";

test("converts AppEvent to bytes", () => {
  const utf8Encoder = new TextEncoder();
  const timestamp = Date.now();
  const name = "eventName";
  const nameBytes = utf8Encoder.encode(name);
  const payload = new Uint8Array([219, 109, 182, 219, 109, 182]);
  const converter = new AppEventBinaryConverter();

  const result = new DataView(converter.toBytes({ timestamp, name, payload }));
  let bufPos = 0;

  // result has correct byte length
  expect(result.byteLength).toEqual(
    EPOCH_MILLISECOND_BYTES +
      NAME_HEADER_BYTES +
      nameBytes.length +
      PAYLOAD_HEADER_BYTES +
      payload.length,
  );

  expect(Number(result.getBigUint64(bufPos))).toEqual(timestamp);
  bufPos += 8;

  // name header equals name byte length
  expect(result.getInt8(bufPos)).toEqual(nameBytes.length);
  bufPos += 1;

  // name bytes equals event name
  for (let i = 0; i < nameBytes.length; i++) {
    expect(result.getUint8(bufPos++)).toEqual(nameBytes[i]);
  }

  // payload header equals payload byte length
  expect(result.getUint32(bufPos)).toEqual(payload.length);
  bufPos += 4;

  // payload bytes equals message payload
  for (let i = 0; i < payload.length; i++) {
    expect(result.getUint8(bufPos++)).toEqual(payload[i]);
  }
});

test("converts event bytes to AppEvent", () => {
  const converter = new AppEventBinaryConverter();
  const timestamp = Date.now();
  const name = "eventName";
  const payload = new Uint8Array([219, 109, 182, 219, 109, 182]);
  const binaryMessage = converter.toBytes({
    timestamp,
    name,
    payload,
  });

  const result = converter.fromBytes(binaryMessage);

  expect(result.timestamp).toEqual(timestamp);
  expect(result.name).toEqual(name);

  const resultPayload = new Uint8Array(result.payload);

  for (let i = 0; i < payload.length; i++) {
    expect(resultPayload[i]).toEqual(payload[i]);
  }
});
